usr/sbin/p11sak/p11sak.c

```diff
@@ -1908,109 +2129,266 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
+        printf("Skip deleting Key\n");
+    } else {
+        printf("Please just enter (y) for yes or (n) for no.\n");
+    }
+
+    return rc;
+}
+
+/**
+ * Delete objects
+ */
+static CK_RV deleteKey(CK_SESSION_HANDLE session, p11sak_kt kt, char *rm_label,
+                       CK_BBOOL *forceAll)
+{
+    CK_ULONG keylength, count;
+    CK_OBJECT_CLASS keyclass;
+    CK_OBJECT_HANDLE hkey;
+    char *keytype = NULL;
+    char *label = NULL;
+    CK_RV rc;
+    char *user_input = NULL;
+
+    rc = tok_key_list_init(session, kt, label);
+    if (rc != CKR_OK) {
+        printf("Init token key list failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
+        return rc;
+    }
+
+    while (1) {
+        rc = funcs->C_FindObjects(session, &hkey, 1, &count);
+        if (rc != CKR_OK) {
+            printf("C_FindObjects failed (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            return 1;
+        }
+        if (count == 0)
+            break;
+
+        rc = tok_key_get_key_type(session, hkey, &keyclass, &keytype,
+                &keylength);
+        if (rc != CKR_OK) {
+            printf("Invalid key type (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            continue;
+        }
+
+        rc = tok_key_get_label_attr(session, hkey, &label);
+        if (rc != CKR_OK) {
+            printf("Retrieval of label failed (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+        }
+
+        if (*forceAll) {
+            if ((strcmp(rm_label, "") == 0) || (strcmp(rm_label, label) == 0)) {
+                printf("Destroy Object with Label: %s\n", label);
+                rc = funcs->C_DestroyObject(session, hkey);
+                if (rc != CKR_OK) {
+                    printf(
+                            "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                            label, rc, p11_get_ckr(rc));
+                    return rc;
+                }
+                printf("DONE - Destroy Object with Label: %s\n", label);
+            }
+        } else {
+            if ((strcmp(rm_label, "") == 0) || (strcmp(rm_label, label) == 0)) {
+                rc = confirmDestroy(&user_input, label);
+                if (rc != CKR_OK) {
+                    printf("User input failed (error code 0x%lX: %s)\n",
+                            rc, p11_get_ckr(rc));
+                    return rc;
+                }
+
+                rc = finalizeDestroyObject(user_input, label, &session, &hkey);
+                if (rc != CKR_OK) {
+                    return rc;
+                }
+            }
+            free(user_input);
+        }
+
+        free(label);
+        free(keytype);
+    }
+
+    rc = funcs->C_FindObjectsFinal(session);
+    if (rc != CKR_OK) {
+        printf("C_FindObjectsFinal failed (error code 0x%lX: %s)\n", rc,
+                p11_get_ckr(rc));
+        goto done;
+    }
+
+    rc = CKR_OK;
+
+done:
```

Comment:
Why not use this exit label for more error exits to free buffers etc ?

Reply:
Made use of the exit label to free buffers.